### ISSUE TITLE
[congruence] is extremely slow with a large number in the context

### ISSUE THREAD

**Issue by Zimmi48** [Zimmi48/bugzilla-test#5548]
Note: the issue was created automatically with bugzilla2github tool

# Bugzilla Bug ID: 5548
Date: 2017-05-22 16:03:37 +0200
From: @_tchajed
To:   
CC:   coq-bugs-redist@lists.gforge.inria.fr, @_cmangin, @_JasonGross

Last updated: 2017-05-23 18:56:12 +0200

## Bugzilla Comment ID: 16384
Date: 2017-05-22 16:03:37 +0200
From: @_tchajed

The following takes a long time (6 seconds on my machine):

Goal forall o, 4000 = 4000 ->
          o = Some tt ->
          o = None ->
          False.
Proof.
  intros.
  time congruence.
Qed.

I understand that 4000 is a large term, but it seems like congruence shouldn't 
be affected since that hypothesis isn't relevant to this proof. Alternately, if 
some part of what congruence does requires it to search subterms, I'd like 
variants of congruence that only do specific things so I can call only the fast 
ones most of the time.

## Bugzilla Comment ID: 16389
Date: 2017-05-22 20:07:14 +0200
From: @_JasonGross

I would've guessed the issue is that [congruence] somewhere looks inside subterms to discover if it can contradict them, but
Goal forall o, 400 = 401 ->
               o = Some tt ->
               (*o = None ->*)
               False.
  intros.
  Time congruence. (* 8 s *)
while
Goal forall o, 400 = 401 ->
               o = Some tt ->
               o = None ->
               False.
  intros.
  Time congruence. (* 0.1 s *)
so it can't just be that...

## Bugzilla Comment ID: 16400
Date: 2017-05-23 11:08:10 +0200
From: @_cmangin

There is a similar time difference between 

Goal forall o, 4000 = 4001 ->
  o = Some tt ->
  o = None ->
  False.
Proof.
  intros.
  Time congruence.
Qed.

and

Goal forall o, 4000 = 4001 ->
  o = Some tt ->
(*  o = None -> *)
  False.
Proof.
  intros.
  Time congruence.
Qed.

(The first one just made Coq crash on my machine, because the term to build is too big.)

In other words, I think it's just that: it takes some time to look into 4000 = 4000 to check for an absurdity, and what took time in the exemple with 400 = 401 is just building the proof term, which is quite big.
You'll notice that the following example takes several seconds too:

Goal forall o, o = Some tt ->
               o = None ->
               400 = 401 ->
               False.
  intros.
  Time congruence.
Qed.

because it chooses to prove the absurdity with "400 = 401" instead of "Some tt = None".

In other

## Bugzilla Comment ID: 16401
Date: 2017-05-23 14:04:11 +0200
From: @_JasonGross

> I think it's just that: it takes some time to look into

Can such a test be protected by a constr_eq-based fail-fast path?

## Bugzilla Comment ID: 16402
Date: 2017-05-23 18:49:52 +0200
From: @_cmangin

After reading a bit of the code, I think I understand where most of the slowness comes from. Before even executing the congruence algorithm, it first deconstructs both sides of any equality to do some preliminary work.

It should be possible to add some heuristics to ignore some equalities, but it wouldn't solve the problem in general. For instance, if you replace "4000 = 4000" by "S ( S ( .. S x)) = S ( S ( .. S y))", then it would still deconstruct these terms before executing the congruence algorithm.

I have yet to look into the issue further to check if it's just a matter of doing this preliminary work more lazily, or if it cannot be avoided, but I would be a bit pessimistic about having a real improvement easily.

On a positive note, I have confirmed that the execution of the algorithm itself is not responsible for the slowness: once the preliminary work is done, it won't do a lot of unnecessary steps. You can use the flag "Congruence Verbose" to check by yourself.
(Technical detail: this is because the algorithm works in a somewhat "breadth-first" fashion, so it will discover the absurdity involving "Some tt" and "None" quickly enough.)

## Bugzilla Comment ID: 16403
Date: 2017-05-23 18:53:39 +0200
From: @_tchajed

I don't think this has anything to do with equalities - that example just 
incidentally used equality to get a number in the context. Here congruence 
can't do anything with the F 4000 term, but it still examines the whole thing 
and takes 6 seconds:

Axiom F: nat -> Type.

Goal forall o, F 4000 ->
          o = Some tt ->
          o = None ->
          False.
Proof.
  intros.
  time congruence.
Qed.

## Bugzilla Comment ID: 16404
Date: 2017-05-23 18:56:12 +0200
From: @_tchajed

[Set Congruence Debug] is awesome, thanks for point that out, Enrico!

It looks like the slowness is actually in building the problem; the pause is 
between these two lines:

Debug: Reading subgoal ...
Debug: Problem built, solving ...

In my real use case I've fixed this issue by wrapping the number in a 
definition, which mean congruence doesn't see it; it seems like that might be 
necessary?
